discojs/discojs-core/src/dataset/data/preprocessing/text_preprocessing.ts

```diff
@@ -1,57 +1,84 @@
 import { List } from 'immutable'
 import * as tf from '@tensorflow/tfjs'
 
+import type { Task } from '../../../index.js'
 import type { PreprocessingFunction } from './base.js'
+import { AutoTokenizer, PreTrainedTokenizer } from '@xenova/transformers';
 
 /**
  * Available text preprocessing types.
  */
 export enum TextPreprocessing {
   Tokenize,
-  Padding
-}
-
-interface TextEntry extends tf.TensorContainerObject {
-  xs: string[]
-  ys: number[]
+  LeftPadding
 }
 
 interface TokenizedEntry extends tf.TensorContainerObject {
   xs: tf.Tensor1D
-  ys: tf.Tensor1D
 }
 
-// TODO that'll fail everytime
-const gpt3Tokenizer = null as unknown as { encode: (_: string) => { bpe: number[]; text: string[] } }
-
-const padding: PreprocessingFunction = {
-  type: TextPreprocessing.Padding,
-  apply: (x: tf.TensorContainer) => {
-    const { xs, ys } = x as TokenizedEntry
-    // TODO: add to task definition
-    const maxLength = 64
-    if (maxLength === undefined) {
-      return { xs, ys }
+/**
+ * We are currently only implementing left padding for text generation 
+ * https://huggingface.co/docs/transformers/en/llm_tutorial#wrong-padding-side
+ * The function can easily be extended to support right padding once the need arise
+ */
+const leftPadding: PreprocessingFunction = {
+  type: TextPreprocessing.LeftPadding,
+  apply: async (x: Promise<tf.TensorContainer>, task: Task): Promise<tf.TensorContainer> => {
+    let { xs } = await x as TokenizedEntry
+    let tokenizer = task.trainingInformation.tokenizer as PreTrainedTokenizer
+    if (tokenizer === undefined) {
+      const tokenizerName = task.trainingInformation.tokenizerName ?? 'Xenova/gpt2'
+      tokenizer = await AutoTokenizer.from_pretrained(tokenizerName)
+      task.trainingInformation.tokenizer = tokenizer
+    }
+    const maxLength = task.trainingInformation.maxSequenceLength ?? tokenizer.model_max_length as number
+    // Should never happen because tokenization truncates inputs
+    if (xs.size > maxLength) {
+      xs = xs.slice([0], [maxLength])
+    } else if (xs.size < maxLength) {
+      const paddingToken = tokenizer.pad_token_id
+      xs = xs.pad([[Math.max(0, maxLength - xs.size), 0]], paddingToken)
```

Comment:
~~note there is [support for padding in transformer.js' tokenizers](https://github.com/xenova/transformers.js/blob/main/src/tokenizers.js#L2498). from what I saw in code, it's not allowing to left pad, only to right pad (still in a TODO state). consider adding a line mention to know where to look for updates.~~

and I read a bit further now, you already thought of that :)